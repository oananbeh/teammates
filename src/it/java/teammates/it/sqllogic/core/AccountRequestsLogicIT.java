package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.AccountRequestStatus;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testGetAccountRequest_nonExistentAccountRequest_returnsNull() {
        UUID id = UUID.randomUUID();
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertNull(actualAccountRequest);
    }

    @Test
    public void testGetAccountRequest_existingAccountRequest_getsSuccessfully() throws InvalidParametersException {
        AccountRequest expectedAccountRequest =
                new AccountRequest("test@gmail.com", "name", "institute", AccountRequestStatus.PENDING, "comments");
        UUID id = expectedAccountRequest.getId();
        accountRequestsLogic.createAccountRequest(expectedAccountRequest);
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertEquals(expectedAccountRequest, actualAccountRequest);
    }

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";
        AccountRequestStatus status = AccountRequestStatus.PENDING;
        String comments = "comments";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute, status, comments);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        UUID id = toReset.getId();
        toReset = accountRequestsDb.getAccountRequest(id);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(id);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(toReset.getId());

        assertNull(accountRequestsLogic.getAccountRequest(toReset.getId()));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(id));
    }
}
