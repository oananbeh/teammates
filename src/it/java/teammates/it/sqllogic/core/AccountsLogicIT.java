package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.NotificationStyle;
import teammates.common.datatransfer.NotificationTargetUser;
import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.FieldValidator;
import teammates.common.util.HibernateUtil;
import teammates.common.util.StringHelper;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountsLogic;
import teammates.sqllogic.core.CoursesLogic;
import teammates.sqllogic.core.NotificationsLogic;
import teammates.sqllogic.core.UsersLogic;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlentity.Account;
import teammates.storage.sqlentity.Course;
import teammates.storage.sqlentity.Instructor;
import teammates.storage.sqlentity.Notification;
import teammates.storage.sqlentity.ReadNotification;
import teammates.storage.sqlentity.Student;
import teammates.test.AssertHelper;

/**
 * SUT: {@link AccountsLogic}.
 */
public class AccountsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountsLogic accountsLogic = AccountsLogic.inst();
    private NotificationsLogic notificationsLogic = NotificationsLogic.inst();
    private UsersLogic usersLogic = UsersLogic.inst();
    private CoursesLogic coursesLogic = CoursesLogic.inst();

    private AccountsDb accountsDb = AccountsDb.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testUpdateReadNotifications()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {
        ______TS("success: mark notification as read");
        Account account = new Account("google-id", "name", "email@teammates.com");
        Notification notification = new Notification(Instant.parse("2011-01-01T00:00:00Z"),
                Instant.parse("2099-01-01T00:00:00Z"), NotificationStyle.DANGER, NotificationTargetUser.GENERAL,
                "A deprecation note", "<p>Deprecation happens in three minutes</p>");
        accountsDb.createAccount(account);
        notificationsLogic.createNotification(notification);

        String googleId = account.getGoogleId();
        UUID notificationId = notification.getId();
        accountsLogic.updateReadNotifications(googleId, notificationId, notification.getEndTime());

        Account actualAccount = accountsDb.getAccountByGoogleId(googleId);
        List<ReadNotification> accountReadNotifications = actualAccount.getReadNotifications();
        assertEquals(1, accountReadNotifications.size());
        assertSame(actualAccount, accountReadNotifications.get(0).getAccount());
        assertSame(notification, accountReadNotifications.get(0).getNotification());
    }

    @Test
    public void testJoinCourseForStudent()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        Student student2YetToJoinCourse = typicalDataBundle.students.get("student2YetToJoinCourse4");
        Student student3YetToJoinCourse = typicalDataBundle.students.get("student3YetToJoinCourse4");
        Student studentInCourse = typicalDataBundle.students.get("student1InCourse1");

        String loggedInGoogleId = "AccLogicT.student.id";

        ______TS("failure: wrong key");

        String wrongKey = StringHelper.encrypt("wrongkey");
        EntityDoesNotExistException ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForStudent(wrongKey, loggedInGoogleId));
        assertEquals("No student with given registration key: " + wrongKey, ednee.getMessage());

        ______TS("failure: invalid parameters");

        InvalidParametersException ipe = assertThrows(InvalidParametersException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), "wrong student"));
        AssertHelper.assertContains(FieldValidator.REASON_INCORRECT_FORMAT, ipe.getMessage());

        ______TS("failure: googleID belongs to an existing student in the course");

        EntityAlreadyExistsException eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(),
                studentInCourse.getGoogleId()));
        assertEquals("Student has already joined course", eaee.getMessage());

        ______TS("success: with encryption and new account to be created");

        accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), loggedInGoogleId);
        Account accountCreated = accountsLogic.getAccountForGoogleId(loggedInGoogleId);

        assertEquals(loggedInGoogleId, usersLogic.getStudentForEmail(
                student2YetToJoinCourse.getCourseId(), student2YetToJoinCourse.getEmail()).getGoogleId());
        assertNotNull(accountCreated);

        ______TS("success: student joined but account already exists");

        String existingAccountId = "existingAccountId";
        Account existingAccount = new Account(existingAccountId, "accountName", student3YetToJoinCourse.getEmail());
        accountsDb.createAccount(existingAccount);

        accountsLogic.joinCourseForStudent(student3YetToJoinCourse.getRegKey(), existingAccountId);

        assertEquals(existingAccountId, usersLogic.getStudentForEmail(
                student3YetToJoinCourse.getCourseId(), student3YetToJoinCourse.getEmail()).getGoogleId());

        ______TS("failure: already joined");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), loggedInGoogleId));
        assertEquals("Student has already joined course", eaee.getMessage());

        ______TS("failure: course is deleted");

        Course originalCourse = usersLogic.getStudentForEmail(
                student2YetToJoinCourse.getCourseId(), student2YetToJoinCourse.getEmail()).getCourse();
        coursesLogic.moveCourseToRecycleBin(originalCourse.getId());

        ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(),
                        loggedInGoogleId));
        assertEquals("The course you are trying to join has been deleted by an instructor", ednee.getMessage());
    }

    @Test
    public void testJoinCourseForInstructor() throws Exception {
        String instructorIdAlreadyJoinedCourse = "instructor1";
        Instructor instructor2YetToJoinCourse = typicalDataBundle.instructors.get("instructor2YetToJoinCourse4");
        Instructor instructor3YetToJoinCourse = typicalDataBundle.instructors.get("instructor3YetToJoinCourse4");

        String loggedInGoogleId = "AccLogicT.instr.id";
        String[] key = new String[] {
                getRegKeyForInstructor(instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail()),
                getRegKeyForInstructor(instructor2YetToJoinCourse.getCourseId(), instructor3YetToJoinCourse.getEmail()),
        };

        ______TS("failure: googleID belongs to an existing instructor in the course");

        EntityAlreadyExistsException eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(
                        key[0], instructorIdAlreadyJoinedCourse));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("success: instructor joined and new account be created");

        accountsLogic.joinCourseForInstructor(key[0], loggedInGoogleId);

        Instructor joinedInstructor = usersLogic.getInstructorForEmail(
                        instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail());
        assertEquals(loggedInGoogleId, joinedInstructor.getGoogleId());

        Account accountCreated = accountsLogic.getAccountForGoogleId(loggedInGoogleId);
        assertNotNull(accountCreated);

        ______TS("success: instructor joined but account already exists");

        String existingAccountId = "existingAccountId";
        Account existingAccount = new Account(existingAccountId, "accountName", instructor3YetToJoinCourse.getEmail());
        accountsDb.createAccount(existingAccount);

        accountsLogic.joinCourseForInstructor(key[1], existingAccount.getGoogleId());

        joinedInstructor = usersLogic.getInstructorForEmail(
                        instructor3YetToJoinCourse.getCourseId(), existingAccount.getEmail());
        assertEquals(existingAccountId, joinedInstructor.getGoogleId());

        ______TS("failure: instructor already joined");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(key[0], loggedInGoogleId));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("failure: key belongs to a different user");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(key[0], "otherUserId"));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("failure: invalid key");

        String invalidKey = StringHelper.encrypt("invalidKey");

        EntityDoesNotExistException ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForInstructor(invalidKey, loggedInGoogleId));
        assertEquals("No instructor with given registration key: " + invalidKey,
                ednee.getMessage());

        ______TS("failure: course deleted");

        Course originalCourse = usersLogic.getInstructorForEmail(
                instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail()).getCourse();
        coursesLogic.moveCourseToRecycleBin(originalCourse.getId());

        ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForInstructor(instructor2YetToJoinCourse.getRegKey(),
                    instructor2YetToJoinCourse.getGoogleId()));
        assertEquals("The course you are trying to join has been deleted by an instructor", ednee.getMessage());
    }

    private String getRegKeyForInstructor(String courseId, String email) {
        return usersLogic.getInstructorForEmail(courseId, email).getRegKey();
    }
}
