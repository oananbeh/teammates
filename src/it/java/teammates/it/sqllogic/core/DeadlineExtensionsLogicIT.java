package teammates.it.sqllogic.core;

import java.time.Instant;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.DeadlineExtensionsLogic;
import teammates.storage.sqlentity.FeedbackSession;
import teammates.storage.sqlentity.Student;

/**
 * SUT: {@link DeadlineExtensionsLogic}.
 */
public class DeadlineExtensionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testGetExtendedDeadline_extensionExists_success() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        assert student != null;
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNotNull(extendedDeadlineForStudent);
        assertEquals(Instant.parse("2027-04-30T23:00:00Z"), extendedDeadlineForStudent);
    }

    @Test
    public void testGetExtendedDeadline_extensionDoesNotExist_null() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student2InCourse1");
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNull(extendedDeadlineForStudent);
    }
}
