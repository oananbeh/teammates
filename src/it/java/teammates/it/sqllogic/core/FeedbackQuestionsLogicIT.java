package teammates.it.sqllogic.core;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.FeedbackParticipantType;
import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.datatransfer.questions.FeedbackQuestionDetails;
import teammates.common.datatransfer.questions.FeedbackQuestionType;
import teammates.common.datatransfer.questions.FeedbackTextQuestionDetails;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.FeedbackQuestionsLogic;
import teammates.storage.sqlentity.FeedbackQuestion;
import teammates.storage.sqlentity.FeedbackSession;
import teammates.ui.output.FeedbackVisibilityType;
import teammates.ui.output.NumberOfEntitiesToGiveFeedbackToSetting;
import teammates.ui.request.FeedbackQuestionUpdateRequest;

/**
 * SUT: {@link FeedbackQuestionsLogic}.
 */
public class FeedbackQuestionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
    }

    @Test
    public void testCreateFeedbackQuestion() throws InvalidParametersException, EntityAlreadyExistsException {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackTextQuestionDetails newQuestionDetails = new FeedbackTextQuestionDetails("New question text.");
        List<FeedbackParticipantType> showTos = new ArrayList<>();
        showTos.add(FeedbackParticipantType.INSTRUCTORS);
        FeedbackQuestion newQuestion = FeedbackQuestion.makeQuestion(fs, 6, "This is a new text question",
                FeedbackParticipantType.STUDENTS, FeedbackParticipantType.OWN_TEAM_MEMBERS, -100,
                showTos, showTos, showTos, newQuestionDetails);

        newQuestion = fqLogic.createFeedbackQuestion(newQuestion);

        FeedbackQuestion actualQuestion = fqLogic.getFeedbackQuestion(newQuestion.getId());

        verifyEquals(newQuestion, actualQuestion);
    }

    @Test
    public void testGetFeedbackQuestionsForSession() {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackQuestion fq1 = typicalDataBundle.feedbackQuestions.get("qn1InSession1InCourse1");
        FeedbackQuestion fq2 = typicalDataBundle.feedbackQuestions.get("qn2InSession1InCourse1");
        FeedbackQuestion fq3 = typicalDataBundle.feedbackQuestions.get("qn3InSession1InCourse1");
        FeedbackQuestion fq4 = typicalDataBundle.feedbackQuestions.get("qn4InSession1InCourse1");
        FeedbackQuestion fq5 = typicalDataBundle.feedbackQuestions.get("qn5InSession1InCourse1");
        FeedbackQuestion fq6 = typicalDataBundle.feedbackQuestions.get("qn6InSession1InCourse1NoResponses");
        FeedbackQuestion fq7 = typicalDataBundle.feedbackQuestions.get("qn7InSession1InCourse1");
        FeedbackQuestion fq8 = typicalDataBundle.feedbackQuestions.get("qn8InSession1InCourse1");
        FeedbackQuestion fq9 = typicalDataBundle.feedbackQuestions.get("qn9InSession1InCourse1");

        List<FeedbackQuestion> expectedQuestions = List.of(fq1, fq2, fq3, fq4, fq5, fq6, fq7, fq8, fq9);

        List<FeedbackQuestion> actualQuestions = fqLogic.getFeedbackQuestionsForSession(fs);

        assertEquals(expectedQuestions.size(), actualQuestions.size());
        assertTrue(expectedQuestions.containsAll(actualQuestions));
    }

    @Test
    public void testUpdateFeedbackQuestionCascade() throws InvalidParametersException, EntityDoesNotExistException {
        FeedbackQuestion fq1 = typicalDataBundle.feedbackQuestions.get("qn1InSession1InCourse1");
        fq1.setDescription("New question description");
        FeedbackQuestionUpdateRequest updateRequest = generateFeedbackQuestionUpdateRequest(
                fq1.getQuestionNumber(),
                fq1.getDescription(),
                fq1.getQuestionDetailsCopy(),
                fq1.getQuestionDetailsCopy().getQuestionType(),
                fq1.getGiverType(),
                fq1.getRecipientType(),
                fq1.getNumOfEntitiesToGiveFeedbackTo(),
                fq1.getShowResponsesTo(),
                fq1.getShowGiverNameTo(),
                fq1.getShowRecipientNameTo()
        );
        updateRequest.setNumberOfEntitiesToGiveFeedbackToSetting(NumberOfEntitiesToGiveFeedbackToSetting.CUSTOM);

        fqLogic.updateFeedbackQuestionCascade(fq1.getId(), updateRequest);

        FeedbackQuestion actualFeedbackQuestion = fqLogic.getFeedbackQuestion(fq1.getId());

        verifyEquals(fq1, actualFeedbackQuestion);
    }

    private FeedbackQuestionUpdateRequest generateFeedbackQuestionUpdateRequest(
            int questionNumber,
            String questionDescription,
            FeedbackQuestionDetails questionDetails,
            FeedbackQuestionType questionType,
            FeedbackParticipantType giverType,
            FeedbackParticipantType recipientType,
            Integer customNumberOfEntitiesToGiveFeedbackTo,
            List<FeedbackParticipantType> showResponsesTo,
            List<FeedbackParticipantType> showGiverNameTo,
            List<FeedbackParticipantType> showRecipientNameTo
    ) {
        FeedbackQuestionUpdateRequest updateRequest = new FeedbackQuestionUpdateRequest();

        updateRequest.setQuestionNumber(questionNumber);
        updateRequest.setQuestionDescription(questionDescription);
        updateRequest.setQuestionDetails(questionDetails);
        updateRequest.setQuestionType(questionType);
        updateRequest.setGiverType(giverType);
        updateRequest.setRecipientType(recipientType);
        updateRequest.setCustomNumberOfEntitiesToGiveFeedbackTo(customNumberOfEntitiesToGiveFeedbackTo);
        updateRequest.setShowResponsesTo(convertToFeedbackVisibilityType(showResponsesTo));
        updateRequest.setShowGiverNameTo(convertToFeedbackVisibilityType(showGiverNameTo));
        updateRequest.setShowRecipientNameTo(convertToFeedbackVisibilityType(showRecipientNameTo));

        return updateRequest;
    }

    private List<FeedbackVisibilityType> convertToFeedbackVisibilityType(
            List<FeedbackParticipantType> feedbackParticipantTypes) {
        return feedbackParticipantTypes.stream().map(feedbackParticipantType -> {
            switch (feedbackParticipantType) {
            case STUDENTS:
                return FeedbackVisibilityType.STUDENTS;
            case INSTRUCTORS:
                return FeedbackVisibilityType.INSTRUCTORS;
            case RECEIVER:
                return FeedbackVisibilityType.RECIPIENT;
            case OWN_TEAM_MEMBERS:
                return FeedbackVisibilityType.GIVER_TEAM_MEMBERS;
            case RECEIVER_TEAM_MEMBERS:
                return FeedbackVisibilityType.RECIPIENT_TEAM_MEMBERS;
            default:
                assert false : "Unknown feedbackParticipantType" + feedbackParticipantType;
                break;
            }
            return null;
        }).collect(Collectors.toList());
    }
}
