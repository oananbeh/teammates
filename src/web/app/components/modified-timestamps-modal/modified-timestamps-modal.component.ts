import { Component, Input } from '@angular/core';
import { TweakedTimestampData } from '../../pages-instructor/instructor-session-base-page.component';

@Component({
  selector: 'tm-modified-timestamps-modal',
  templateUrl: './modified-timestamps-modal.component.html',
  styleUrls: ['./modified-timestamps-modal.component.scss'],
})
export class ModifiedTimestampModalComponent {
  @Input()
  coursesOfModifiedSession: string[] = [];
  @Input()
  modifiedSessions: Record<string, TweakedTimestampData> = {};
}
