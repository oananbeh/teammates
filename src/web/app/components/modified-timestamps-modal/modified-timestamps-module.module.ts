import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ModifiedTimestampModalComponent } from './modified-timestamps-modal.component';

/**
 * Module for instructor home page.
 */
@NgModule({
  declarations: [
    ModifiedTimestampModalComponent,
  ],
  imports: [
    CommonModule,
  ],
  exports: [
    ModifiedTimestampModalComponent,
  ],
})
export class ModifiedTimestampModalModule { }
