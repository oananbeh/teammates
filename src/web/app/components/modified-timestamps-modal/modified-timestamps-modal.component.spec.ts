import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ModifiedTimestampModalComponent } from './modified-timestamps-modal.component';
import { TweakedTimestampData } from '../../pages-instructor/instructor-session-base-page.component';

describe('ModifiedTimestampModalComponent', () => {
  let component: ModifiedTimestampModalComponent;
  let fixture: ComponentFixture<ModifiedTimestampModalComponent>;

  const coursesOfModifiedSession: string[] = [
    'CS315-Semester12023',
  ];

  const modifiedSessions: Record<string, TweakedTimestampData> = {
    'Instructor Feedback Session': {
      newTimestamp: {
        submissionStartTimestamp: '23 Apr 2023 9:00 PM',
        submissionEndTimestamp: '27 Apr 2023 11:59 PM',
        sessionVisibleTimestamp: 'On submission opening time',
        responseVisibleTimestamp: 'Not now (publish manually)',
      },
      oldTimestamp: {
        submissionStartTimestamp: '22 Mar 2023 9:00 PM',
        submissionEndTimestamp: '31 Mar 2023 11:59 PM',
        sessionVisibleTimestamp: 'On submission opening time',
        responseVisibleTimestamp: 'Not now (publish manually)',
      },
    },
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ModifiedTimestampModalComponent],
    })
    .compileComponents();

    fixture = TestBed.createComponent(ModifiedTimestampModalComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with coursesOfModifiedSession and modifiedSessions', () => {
    component.coursesOfModifiedSession = coursesOfModifiedSession;
    component.modifiedSessions = modifiedSessions;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should snap with modifiedSessions', () => {
    component.modifiedSessions = modifiedSessions;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

});
