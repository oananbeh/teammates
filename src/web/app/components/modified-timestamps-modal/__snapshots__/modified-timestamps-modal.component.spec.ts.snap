// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModifiedTimestampModalComponent should snap with coursesOfModifiedSession and modifiedSessions 1`] = `
<tm-modified-timestamps-modal
  coursesOfModifiedSession={[Function Array]}
  modifiedSessions={[Function Object]}
>
  <p>
     The feedback session has been copied to all courses. However, changes are made to some session timings as the existing timings are no longer valid in these courses: CS315-Semester12023. The changes are as follows:
  </p><table
    class="table table-bordered"
  >
    <thead>
      <tr>
        <th>
          Session Name
        </th>
        <th>
          Timings
        </th>
        <th>
          Existing
        </th>
        <th>
          Modified
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          rowspan="4"
        >
          Instructor Feedback Session
        </td>
        <td>
          Submission opening time
        </td>
        <td>
          22 Mar 2023 9:00 PM
        </td>
        <td>
          23 Apr 2023 9:00 PM
        </td>
      </tr>
      <tr>
        <td>
          Submission closing time
        </td>
        <td>
          31 Mar 2023 11:59 PM
        </td>
        <td>
          27 Apr 2023 11:59 PM
        </td>
      </tr>
      <tr
        class="text-color-lightgray"
      >
        <td>
          Session visible time
        </td>
        <td>
          On submission opening time
        </td>
        <td>
          On submission opening time
        </td>
      </tr>
      <tr
        class="text-color-lightgray"
      >
        <td>
          Response visible time
        </td>
        <td>
          Not now (publish manually)
        </td>
        <td>
          Not now (publish manually)
        </td>
      </tr>
    </tbody>
  </table><br /><p>
     Please modify the timings as necessary.
  </p>
</tm-modified-timestamps-modal>
`;

exports[`ModifiedTimestampModalComponent should snap with default fields 1`] = `
<tm-modified-timestamps-modal
  coursesOfModifiedSession={[Function Array]}
  modifiedSessions={[Function Object]}
>
  <p>
     The course has been added. However, changes are made to some session timings as the existing timings are no longer valid. The changes are as follows:
  </p><table
    class="table table-bordered"
  >
    <thead>
      <tr>
        <th>
          Session Name
        </th>
        <th>
          Timings
        </th>
        <th>
          Existing
        </th>
        <th>
          Modified
        </th>
      </tr>
    </thead>
    <tbody>
    </tbody>
  </table><br /><p>
     Please modify the timings as necessary.
  </p>
</tm-modified-timestamps-modal>
`;

exports[`ModifiedTimestampModalComponent should snap with modifiedSessions 1`] = `
<tm-modified-timestamps-modal
  coursesOfModifiedSession={[Function Array]}
  modifiedSessions={[Function Object]}
>
  <p>
     The course has been added. However, changes are made to some session timings as the existing timings are no longer valid. The changes are as follows:
  </p><table
    class="table table-bordered"
  >
    <thead>
      <tr>
        <th>
          Session Name
        </th>
        <th>
          Timings
        </th>
        <th>
          Existing
        </th>
        <th>
          Modified
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          rowspan="4"
        >
          Instructor Feedback Session
        </td>
        <td>
          Submission opening time
        </td>
        <td>
          22 Mar 2023 9:00 PM
        </td>
        <td>
          23 Apr 2023 9:00 PM
        </td>
      </tr>
      <tr>
        <td>
          Submission closing time
        </td>
        <td>
          31 Mar 2023 11:59 PM
        </td>
        <td>
          27 Apr 2023 11:59 PM
        </td>
      </tr>
      <tr
        class="text-color-lightgray"
      >
        <td>
          Session visible time
        </td>
        <td>
          On submission opening time
        </td>
        <td>
          On submission opening time
        </td>
      </tr>
      <tr
        class="text-color-lightgray"
      >
        <td>
          Response visible time
        </td>
        <td>
          Not now (publish manually)
        </td>
        <td>
          Not now (publish manually)
        </td>
      </tr>
    </tbody>
  </table><br /><p>
     Please modify the timings as necessary.
  </p>
</tm-modified-timestamps-modal>
`;
